src/services/tokenService.ts

```diff
@@ -1,48 +1,227 @@
+    maxAge: maxAge || generalConfig.jwtTTL,
+    path: '/' // Ensure the cookie is accessible for your entire app
+  });
+};
+
+// Generate a new CSRF token the session
+export const generateCSRFToken = async (cache: Cache): Promise<string> => {
+  try {
+    const csrfToken = uuidv4().replace(/-/g, '').slice(0, generalConfig.csrfLength);
+    const hashedToken = await hashToken(csrfToken);
+
+    // Add the refresh token to the Cache
+    await cache.adapter.set(`csrf:${csrfToken}`, hashedToken, { ttl: generalConfig.csrfTTL });
+    return csrfToken;
+  } catch(err) {
+    formatLogMessage(logger).error(err, 'generateCSRFToken error!');
+    return null;
+  }
+}
+
+// Generate an access token for the User
+const generateAccessToken = (jti: string, user: User): string => {
+  try {
+    const payload: JWTAccessToken = {
       id: user.id,
       email: user.email,
       givenName: user.givenName,
       surName: user.surName,
       affiliationId: user.affiliationId,
       role: user.role.toString() || UserRole.RESEARCHER,
+      jti,
+      expiresIn: generalConfig.jwtTTL,
+    };
+
+    return jwt.sign(payload, generalConfig.jwtSecret as string, { expiresIn: generalConfig.jwtTTL });
+  } catch(err) {
+    if (logger) {
+      formatLogMessage(logger).error(err, `generateAccessToken error - ${err.message}`);
     }
+    throw AuthenticationError(`${DEFAULT_UNAUTHORIZED_MESSAGE} - ${err.message}`);
+  }
+}
+
+// Generate a refresh token for the User and add it to the Cache.
+const generateRefreshToken = async (cache: Cache, jti: string, userId: number): Promise<string> => {
+  try {
+    const payload: JWTRefreshToken = {
+      jti,
+      id: userId,
+      expiresIn: generalConfig.jwtRefreshTTL,
+    };
+
+    const token = jwt.sign(payload, generalConfig.jwtRefreshSecret, { expiresIn: generalConfig.jwtRefreshTTL });
+    const hashedToken = await hashToken(token);
+    // Add the refresh token to the Cache
+    await cache.adapter.set(`dmspr:${jti}`, hashedToken, { ttl: generalConfig.jwtRefreshTTL })
+    return token;
+  } catch(err) {
+    if (logger) {
+      formatLogMessage(logger).error(err, `generateRefreshToken error - ${err.message}`);
+    }
+    throw AuthenticationError(`${DEFAULT_UNAUTHORIZED_MESSAGE} - ${err.message}`);
+  }
+}
+
+// Generate an Access Token and a Refresh Token
+export const generateAuthTokens = async (cache: Cache, user: User): Promise<{ accessToken: string; refreshToken: string }> => {
+  if (generalConfig.jwtSecret && generalConfig.jwtRefreshSecret && user && user.id && user.email) {
     try {
-      return jwt.sign(payload, generalConfig.jwtSecret as string, { expiresIn: generalConfig.jwtTtl });
+      // Generate a unique id for the JWT
+      const jti = new Date().getTime().toString();
+      // Generate an Access Token
+      const accessToken = generateAccessToken(jti, user);
+      // Generate a Refresh Token
+      const refreshToken = await generateRefreshToken(cache, jti, user.id);
```

Comment:
L114 - Does the refresh token have to be refreshed every time a new access token is generated?

Reply:
I guess it depends. Do we want people to be logged in for 24 hours from the time that they sign in and then  have to sign in again or do we want to keep refreshing that token so that the countdown starts again?